core/controller/src/main/scala/org/apache/openwhisk/core/controller/Packages.scala

```diff
@@ -243,7 +243,7 @@ trait WhiskPackagesApi extends WhiskCollectionAPI with ReferencedEntities {
       wp.binding map {
         // pre-existing entity is a binding, check that new binding is valid
         b =>
-          checkBinding(b.fullyQualifiedName)
+          checkBinding(binding.fullyQualifiedName)
```

Comment:
this was previously checking the _existing_ binding, and not the _incoming_ binding from `content.binding`

Reply:
This also suggests that we need a test along these lines:

```
wsk package create p
wsk package bind p b
wsk package delete p
wsk package create p2
wsk package update p2 b # will fail without your patch, and should succeed with
```

Reply:
the error message for fetching a binding missing its references package is confusing - we should open an issue to also fix that.

currently returns: `The requested resource 'guest/b' does not exist.` instead of `Binding references a package that does not exist.`

Reply:
@rabbah I don't see how the last line (`wsk package update p2 b`) will succeed currently. As `b` is a binding already, and updating a binding to another binding was already not allowed

#### [`Packages.scala:L206`](https://github.com/apache/incubator-openwhisk/blob/master/core/controller/src/main/scala/org/apache/openwhisk/core/controller/Packages.scala#L206-L208)
```
// trying to create a new package binding that refers to another binding
case provider if provider.binding.nonEmpty =>
  Future.failed(RejectRequest(BadRequest, Messages.bindingCannotReferenceBinding))
```
While attempting to run your suggested test, I noticed that the command `wsk update p2 b` isn't accepted, and I'm wondering whether you intended to use `wsk bind p2 b` or another operation? Since `p2` is created as a package it shouldn't allow it to be updated as a binding (from my understanding). My apologies if my inference is incorrect here.
```
$ wsk -i package create p
ok: created package p
$ wsk -i package bind p b
ok: created binding b
$ wsk -i package delete p
ok: deleted package p
$ wsk -i package create p2
ok: created package p2
$ wsk -i package update p2 b
error: Invalid argument(s): b. A package name is required.
Run 'wsk --help' for usage.
$ wsk -i package bind p2 b  
error: Binding creation failed: resource already exists (code MjyhYzbZlRWsxjpmNToG4enhbK1j6EYz)
```
> the error message for fetching a binding missing its references package is confusing - we should open an issue to also fix that.

Do you mean that after p is deleted, attempting to fetch `b` is failing with that error?

Reply:
i left out a detail - i edited the cli to send `?overwrite=true` on the `POST`.

Reply:
i left out a detail - i edited the cli to send ?overwrite=true on the POST.
in other words, if this is a REST API test, you can test it.

Reply:
Did you end up being able to bind a new package to an existing binding? I was unable to do so in my original tests. The logic in `L206` would seem to indicate it would prevent that

Reply:
@rabbah note that your example (with `?overwrite=true`) returns 400 as expected today as pointed out in the [source](https://github.com/apache/incubator-openwhisk/blob/master/core/controller/src/main/scala/org/apache/openwhisk/core/controller/Packages.scala#L206-L208).
```
PUT /api/v1/namespaces/guest/packages/b?overwrite=true HTTP/1.1
Content-Type: application/json

{"binding":{"namespace":"guest","name":"p2"}}
```
```
HTTP/1.1 400 Bad Request
Content-Type: application/json

{"error":"Binding references a package that does not exist.","code":"HWkcauMms0QCAFNs4mhDCrtx7EbqHBWF"}
```
 Do you still think there is another failure case we should resolve in this PR?

Reply:
@asteed the error "Binding references a package that does not exist" is wrong since `p2` does exist. The controller tried to look up `p` which was deleted and hence the reason the error is wrong. If the error contained the missing package, it would be: 

```bash
> wsk package bind p2 b
error: Binding references a package 'p' that does not exist.
```

which is nonsensical since we're requesting to bind `p2` and `p` is no longer relevant.

Reply:
@rabbah Thanks, I didn't catch the incorrect error originally, as I took your comments to suggest another area where we could encounter the looping scenario.

Do you think this should be addressed in this PR or filed as a separate bug similar to (or included in) #4130 ?

Reply:
i think your patch may fix/hide this bug.
i'm ok with a seperate pr if the test can be written against the new code.